specifyweb/workbench/upload/treerecord.py

```diff
@@ -55,20 +173,105 @@ def disambiguate(self, disambiguation: DA) -> "ScopedTreeRecord":
+        return set(tree_def_model.objects.filter(id__in=treedefids))
+    
+    def _get_not_null_ranks_columns_in_row(self, row: Row) -> List[TreeRankCell]:
+        # Get rank columns that are not null in the row
+        ranks_columns_in_row_not_null = []
+        for row_key, row_value in row.items():
+            if not row_value:
+                continue
+            for tree_rank_record, cols in self.ranks.items():
+                for col_name, col_opts in cols.items():
+                    formatted_column = f'{col_opts.column} - {col_name}' if col_name != 'name' else col_opts.column
+                    if formatted_column == row_key:
+                        ranks_columns_in_row_not_null.append(
+                            TreeRankCell(
+                                tree_rank_record.treedef_id,  # treedefid
+                                tree_rank_record.rank_name,  # treedefitem_name
+                                col_name,  # tree_node_attribute
+                                row_value,  # upload_value
+                            )
+                        )
+                        break
+        return ranks_columns_in_row_not_null
+    
+    def _filter_target_rank_columns(self, ranks_columns_in_row_not_null, target_rank_treedef_id) -> List[TreeRankCell]:
+        # Filter ranks_columns_in_row_not_null to only include columns that are part of the target treedef
+        return list(
+            filter(
+                lambda rank_column: rank_column.treedef_id == target_rank_treedef_id,
+                ranks_columns_in_row_not_null,
+            )
+        )
+    
+    def rescope_tree_from_row(self, row: Row) -> Tuple["ScopedTreeRecord", Optional["WorkBenchParseFailure"]]:
+        # Starting out with a row with tree data.
+        # The row can have multiple tree ranks, and each rank can be associated with multiple columns, like name, authoer, etc.
+        # First need to match each rank column to a treedefitem.
+        # Second need to narrow down to the target rank based on which values in the row are not null.
+        # Lastly, need to return a new ScopedTreeRecord with the correct treedef, treedefitems, and root for this row.
+
+        tree_def_model = get_treedef_model(self.name)
+        tree_rank_model = get_treedefitem_model(self.name)
+        tree_node_model = getattr(models, self.name.lower().title())
+
+        if len(set(tr.treedef_id for tr in self.ranks.keys())) == 1:
+            return self, None
+
+        # Get rank columns that are not null in the row
+        ranks_columns_in_row_not_null: List[TreeRankCell] = self._get_not_null_ranks_columns_in_row(row)
+
+        # Determine the target treedef based on the columns that are not null
+        targeted_treedefids = set([rank_column.treedef_id for rank_column in ranks_columns_in_row_not_null])
+        if targeted_treedefids is None or len(targeted_treedefids) == 0:
```

Comment:
Why will `targeted_treedefids` be None? Remove of that check, or are you defending against python too 😅?